### PR TITLE
SSH Config Parser and SSH connection support

### PR DESCRIPTION
**Pull request by JeremySkinner**
Hi @dahlbyk I'd appreciate your feedback on this one when you get some time. 

This adds a parser for openssh config files as well as functions for manipulating the connections. The main function is `Connect-Ssh`. If specified with no arguments, it displays a list of available connections:

image

You can also pass in an alias directly (eg `Connect-Ssh Server1`) 

The `Parse-SshConfig` function is a port of https://github.com/dotnil/ssh-config which uses a lot of anonymous local functions to avoid needing recursion. I've replicated the same behaviour using scriptblocks, but that seems slightly unconventional in the powershell world. This function returns an `SshConfig` object that represents the structure of the openssh config file.

The config object can either be used to return raw `ConfigNode` instances that represent entries in the file, or by calling `Compute($hostname)` it'll return a hashtable with all the properties for a connection, taking into account any wildcard settings and applying them.

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,2 +1 @@
-# Enable LF normalization for all files
-* text=auto
```
Comment (dahlbyk) [inline .gitattributes:2]: Why remove? I'm guessing for the test config files?

File: src/OpenSsh.ps1 [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        [string]
+        $SshHost,
+
+        [Parameter()]
+        [string]
+        $Path = (Get-SshPath 'config'),
+
+        [Parameter()]
+        [switch]
+        $Raw
```
Comment (dahlbyk) [inline src/OpenSsh.ps1:28]: > The config object can either be used to return raw `ConfigNode` instances that represent entries in the file, or by calling `Compute($hostname)` it'll return a hashtable with all the properties for a connection, taking into account any wildcard settings and applying them.

Even after reading this, I feel like I'm being dense...but I don't understand what `-Raw` is supposed to do. Maybe documenting it here would help me?

**Review by dahlbyk — COMMENTED** [review]
Sorry for the delay, I was out of town for several days.

File: .gitattributes (hunk 1 above)
Comment (JeremySkinner) [reply .gitattributes:2]: Yes

File: src/OpenSsh.ps1 (hunk 2 above)
Comment (JeremySkinner) [reply src/OpenSsh.ps1:28]: Sure, I can add some notes for that. With Raw set to true, it returns the underlying raw nodes represented in the config as-is, ie essentially a collection of keys, values and separators. 

With raw set to false, it'll use the computed data. Ie, it'll merge wildcard nodes with actual nodes. So if you have this:

```
Host foo
   HostName foo.com

Host *
  IdentityFile /path/to/id_rsa.pub
```

Raw would give you back 2 root nodes (Host foo, with a childnode for the Hostname,  and Host * with a child node for the IdentityFile). 
Without Raw, you'd get a single result for  Host foo, with both a hostname and an identityfile (as the wildcard was expanded).

This mimics the behaviour of the library I ported this from, which supports working with both the underlying raw data as well as the 'computed' entries which will expand wildcards.

I pushed the current state of this out as 0.2 in the end as I needed to start working with it, but can do a patch release if you want any changes.

File: .gitattributes (hunk 1 above)
Comment (dahlbyk) [reply .gitattributes:2]: K, best to put this back then and add `test/fixtures/.gitattributes` with `* -text`.

File: src/OpenSsh.ps1 (hunk 2 above)
Comment (dahlbyk) [reply src/OpenSsh.ps1:28]: It's fine, just trying to understand for review's sake. Don't let me slow you down on this at all. 😀

File: .gitattributes (hunk 1 above)
Comment (JeremySkinner) [reply .gitattributes:2]: Thanks, that did the trick. I think I misunderstood the purpose of `* text=auto`. I thought it forced text files to always be LF, so then it confused me greatly when checking out the test config files converted them to CRLF (even though I'd committed them with LF and always run with autocrlf disabled). I didn't realise you could then override this with a second .gitattributes in a subdir.

**Pull request merged**